### ISSUE TITLE
doesn't correctly parse Exchange 2007 ics

### ISSUE THREAD

**Issue by scanlanjames** [icalendar/icalendar#161]
``` ruby
require 'icalendar'
include Icalendar
test=<<EOF
BEGIN:VCALENDAR
METHOD:COUNTER
PRODID:Microsoft Exchange Server 2007
VERSION:2.0
BEGIN:VTIMEZONE
TZID:Pacific Standard Time
BEGIN:STANDARD
DTSTART:16010101T020000
TZOFFSETFROM:-0700
TZOFFSETTO:-0800
RRULE:FREQ=YEARLY;INTERVAL=1;BYDAY=1SU;BYMONTH=11
END:STANDARD
BEGIN:DAYLIGHT
DTSTART:16010101T020000
TZOFFSETFROM:-0800
TZOFFSETTO:-0700
RRULE:FREQ=YEARLY;INTERVAL=1;BYDAY=2SU;BYMONTH=3
END:DAYLIGHT
END:VTIMEZONE
BEGIN:VEVENT
ATTENDEE;SENT-BY="MAILTO:B4@acme.com";PARTSTAT=TENTATIVE;CN="GG.":MAILTO:GG@acme.com
DTSTART;TZID=Pacific Standard Time:20160713T093000
DTEND;TZID=Pacific Standard Time:20160713T100000
X-MS-OLK-ORIGINALSTART;TZID=Pacific Standard Time:20160713T140000
X-MS-OLK-ORIGINALEND;TZID=Pacific Standard Time:20160713T143000
SUMMARY;LANGUAGE=en-US:New Time Proposed: [External]  intro to SyncUp
LOCATION;LANGUAGE=en-US:
UID:someuuid@syncup.at
SEQUENCE:5
DTSTAMP:20160712T153832Z
COMMENT;LANGUAGE=en-US:

 ------------------
 New Meeting Time Proposed:
 Wednesday, July 13, 2016 9:30 AM-10:00 AM (UTC-08:00) Pacific Time (US &
  Canada).

END:VEVENT
END:VCALENDAR
EOF

Icalendar::Calendar.parse(test).first.events.first.dtstart
```

yields `Wed, 13 Jul 2016 09:30:00 +0000`; it didn't pick up the timezone.

**Comment by scanlanjames** [comment]
it's even the wrong timezone. it's pacific daylight time now.

**Comment by rahearn** [comment]
Yes, right now we're only attempting to do anything with timezones if both:

1) `ActiveSupport` is available
2) The `TZID` matches a timezone returned by `ActiveSupport::TimeZone.all`

That's not true for `Pacific Standard Time`

**Comment by schnika** [comment]
In order to solve this issue, we need to take the VTIMEZONE component into account when building timezone dependent dates (e.g dtstart, dtend, ...)

# Problem

As mentioned by @rahearn only timezones known by `ActiveSupport` are supported.
In the above example the calendar defines a `VTIMEZONE` with a standard and daylight offset. 

The `VEVENT` references the defined timezone `Pacific Standard Time` through the `TZID` where necessary.  

Since `Pacific Standard Time` is not known by `ActiveSupport` the timezone `UTC` is automatically applied and leads to wrong datetime/time calculations

# Proposal

In order to set datetime fields correctly we need to take  the`VTIMEZONE` objects into account when calculating the offset.

So `DTSTART;TZID=Pacific Standard Time:20160713T093000`

would result in `20160713T093000-0800` or `20160713T093000-0700` (depending on the standard/daylight period) instead of `20160713T093000+0000`

# Technical Draft

Before submitting any code I would like to briefly discuss the changes we need to make to the codebase in order to get on the same page.

Please consider `dtstart` as a placeholder for all possible properties with a `tzid` definition.

First of all we need to keep track of the timezones defined in the parsed ics. Otherwise we cannot apply the correct offset to the property.

Secondly I would pass the already calculated offset as a `param` in addition to the `tzid` to the `Icalendar::Values::DateTime`/`Icalendar::Values::Time` builder (called via `klazz.new` in `Icalendar::Parser#wrap_property_value`). When the offset is given apply it to the given date and return the value (don't care about the `tzid`). If the offset is not given, do the 'old' stuff... (find timezone in active support)

In order to get the offset we need to search for a defined timezone component with a name matching the `dtstart tzid`.

If we have a match. apply the correct offset by checking the `STANDARD`/`DAYLIGHT` rrules (this will be the hardest part, I guess) and pass it to `klazz.new` as a param (as described above)

**Comment by espen** [comment]
Timezones are fun. According to [this blog post](http://icalevents.com/2613-what-every-developer-ought-to-know-about-timezones-wrt-calendar-files/) all of the examples below are valid:

> TZID:America/New_York
> DTSTART;TZID=”America/New_York”:20080807T090000
> X-WR-TIMEZONE:Europe/London
> TZID:Europe/London
> DTSTART;TZID=Europe/London:20060707T130000
> TZID:Canberra\, Melbourne\, Sydney
> DTSTART;TZID=”Canberra, Melbourne, Sydney”:20100831T210000
> X-WR-TIMEZONE;VALUE=TEXT:Europe/Oslo
> DTSTART;TZID=”(GMT+01.00) Amsterdam / Berlin / Bern / Rome / Stockholm / Vienna”:20100831T16000
> TZID:(GMT+02.00) Harare/Pretoria
> DTSTART;TZID=US/Central:20080726T190000
> DTSTART;TZID=/mozilla.org/20050126_1/America/Cancun:20070101T090000
> DTSTART;TZID=”(GMT) Greenwich Mean Time – Dublin / Edinburgh / Lisbon /
> TZID:/mozilla.org/20050126_1/Pacific/Yap

**Comment by rahearn** [comment]
@schnika that sounds good. We probably would want a new initializer parameter for the offset, right? Or, if we lump it in with the current params, we'd have to be sure to not output it when calling `to_ical` later on.

**Comment by rahearn** [comment]
This should be addressed by #191
